addons/udes_stock/models/stock_picking_batch.py

```diff
@@ -55,6 +87,159 @@ def _compute_priority(self):
+            ready_picks, other_picks = \
+                self._calculate_pick_groups(self.picking_ids)
+
+        # This transitions should be handled already in _compute_state
+        # but as a guard against it being called else where by someone
+        # lets add it in here as well
+        if not ready_picks and not other_picks:
+            # All picks are complete so we're done
+            self.state = 'done'
+            return True
+
+        elif ready_picks and not other_picks and not self.user_id:
+            # User is removed lets go back to being ready
+            self.state = 'ready'
+            return True
+
+        # None really a transition but a bit of batch pick managemnet
+        # which has to be delt with
+        elif other_picks:
+            # We shouldn't have these check if we can make them ready
+            # otherwise remove them
+            return self._remove_unready_picks()
+
+    def _from_waiting_transitions(self, other_picks=None):
+        """ Transition from waiting to another state"""
+        self.ensure_one()
+
+        # Make sure we have the pick groups
+        if other_picks is None:
+            _, other_picks = self._calculate_pick_groups(self.picking_ids)
+
+        # the check that there is picks should be done already in
+        # _compute_state but as a guard against it being called else
+        # where by someone lets add it in here as well
+        if self.picking_ids and other_picks:
+            # Not all picks are ready don't transition
+            return False
+
+        if self.user_id:
+            self.state = 'in_progress'
+        else:
+            self.state = 'ready'
+        return True
+
+    def _from_ready_transitions(self, other_picks=None):
+        """ Transition from ready to another state"""
+        self.ensure_one()
+
+        # Make sure we have the pick groups
+        if other_picks is None:
+            _, other_picks = self._calculate_pick_groups(self.picking_ids)
+
+        if other_picks:
+            self.state = 'waiting'
+            return True
+
+        elif self.user_id:
+            self.state = 'in_progress'
+            return True
+
+        return False
+
+    def _calculate_pick_groups(self, picks):
+        """Collect picks into groups based on state - complete picks are ignored
+           Groups are
+                ready    : pick.state is assigned
+                other    : pick.state is in any other state than the ones above
+        """
+        Picking = self.env['stock.picking']
+
+        ready_picks = Picking.browse()
+        other_picks = Picking.browse()
+
+        for pick in picks:
+
+            if pick.state == 'assigned':
+                ready_picks |= pick
+
+            elif pick.state in ['done', 'cancel']:
+                continue
+
+            else:
+                other_picks |= pick
+
+        return ready_picks, other_picks
+
+    @api.multi
+    def _remove_unready_picks(self, picks=None):
```

Comment:
In general, optional args are confusing as they imply a default which may not be clear to the user.